webapp/src/main/webapp/src/app/groups/groups.component.ts

```diff
@@ -1,85 +1,109 @@
+  private _modalSubscriptions: Subscription[] = [];
 
   constructor(private route: ActivatedRoute,
               private router: Router,
               private service: GroupService,
               private modalService: BsModalService) {
   }
 
+  get groups(): Group[] {
+    return this._groups;
+  }
+
+  set groups(groups: Group[]) {
+    this._groups = groups;
+    this.updateFilteredGroups();
+  }
+
   ngOnInit() {
-    this.filterOptions = this.route.snapshot.data[ "filterOptions" ];
+    this.filterOptions = this.route.snapshot.data[ 'filterOptions' ];
     this.query = new GroupQuery(this.filterOptions.subjects);
 
-    if(this.filterOptions.schools.length == 0)
+    if (this.filterOptions.schools.length == 0) {
       return;
+    }
 
-    this.schoolDropdownOptions = this.filterOptions.schools.map(school => {
-      return {
-        label: school.name,
-        value: school
-      };
+    this.schoolDropdownOptions = this.filterOptions.schools.map(school => <any>{
+      label: school.name,
+      value: school
     });
 
-    this.route.params.subscribe(p => {
-      let params:any = p;
-
+    this.route.params.subscribe((params: any) => {
       this.query.school = this.filterOptions.schools.find(school => school.id == params.schoolId) || this.filterOptions.schools[ 0 ];
       this.query.schoolYear = this.filterOptions.schoolYears.find(year => year === +params.schoolYear) || this.filterOptions.schoolYears[ 0 ];
-      this.query.subject =  this.filterOptions.subjects.find(subject => subject === params.subject) || this.filterOptions.subjects[ 0 ];
-
+      this.query.subject = this.filterOptions.subjects.find(subject => subject === params.subject) || this.filterOptions.subjects[ 0 ];
       this.updateResults();
     });
   }
 
+  ngOnDestroy() {
+    this.unsubscribe();
+  }
+
+  unsubscribe() {
+    this._modalSubscriptions.forEach(subscription => subscription.unsubscribe());
+    this._modalSubscriptions = [];
+  }
+
   updateRoute() {
     let params = {
       schoolId: this.query.school.id,
       schoolYear: this.query.schoolYear,
       subject: this.query.subject
     };
-
     this.router.navigate([ 'groups', params ]);
   }
 
+  onSearchChange() {
+    this.updateFilteredGroups();
+  }
+
   updateResults() {
     this.service
       .getGroups(this.query)
       .subscribe(groups => {
         this.groups = groups;
-        this.filterGroups();
-      })
+      });
   }
 
-  filterGroups() {
-    this.filteredGroups = this.groups.filter( x => x.name.toUpperCase().indexOf(this.searchTerm.toUpperCase()) >= 0)
+  updateFilteredGroups() {
+    this.filteredGroups = this.groups
+      .filter(x => x.name.toUpperCase().indexOf(this.searchTerm.toUpperCase()) >= 0);
   }
 
-  openSetActiveModal(group: Group) {
-    this.bsModalRef = this.modalService.show(DeleteGroupModalComponent);
-    this.bsModalRef.content.group = group;
-    this.modalService.onHide.subscribe(x => {
-      this.updateResults();
-    });
```

Comment:
This was the root of the issue and was problematic for multiple reasons.
1) we need to clean up the subscription otherwise it will add a new subscription and callback every opening of the same modal.
2) the modal hiding doesn't exactly equate to a deletion having been made, so the group list was updating even when the modal was cancelled
3) we know which group is deleted successfully so we don't need to ping the backend for the reports we can just remove the one we deleted from the list

Reply:
For 2 and 3, my thoughts were just to always refresh the data in case something changed while that modal was up.  For #1, I think we should just unsubscribe on the OnDestroy maybe?

Reply:
Every time subscribe was called on the modalService it was accumulating. I think under the hood when we subscribe to the modal service it holds the event emitter at the service level and doesn't just delegate to the modal instance emitters.

For #2/3 i think if we lookup recent status of the groups when they happen to delete them, they might think they deleted more than just what they selected no?